### PR TITLE
(CDAP-3387) Store the stream format specification as part of SerDe property

### PR DESCRIPTION
**Pull request by chtyim**
- Detach it from StreamConfig so that different stream table view can have different format spec
- Modify the InputFormat/RecordRecord to use the FormatSpecification from the SerDe instead of reading from StreamConfig.

### CODE REVIEW

File: cdap-explore/src/main/java/co/cask/cdap/explore/service/ExploreTableManager.java [hunk 1]
```diff
@@ -188,7 +196,8 @@ public QueryHandle enableDataset(Id.DatasetInstance datasetID, DatasetSpecificat
           ((RecordScannable) dataset).getRecordType() : ((RecordWritable) dataset).getRecordType();
 
         // if the type is a structured record, use the schema property to create the table
-        if (StructuredRecord.class.equals(recordType)) {
+        // Use == because that's what same class means.
+        if (StructuredRecord.class == recordType) {
```
Comment (jwang47) [inline cdap-explore/src/main/java/co/cask/cdap/explore/service/ExploreTableManager.java]: Why does `recordType` have to be == StructuredRecord.class instead of allowing `recordType` to be a subclass?

File: cdap-explore/src/main/java/co/cask/cdap/hive/stream/HiveStreamInputFormat.java [hunk 2]
```diff
@@ -96,7 +93,18 @@ public InputSplit createSplit(Path path, Path indexPath, long startTime, long en
                                                                StreamUtils.getGeneration(streamConfig));
 
     StreamInputSplitFinder.Builder builder = StreamInputSplitFinder.builder(streamPath.toURI());
-    return setupBuilder(conf, streamConfig, builder).build(SPLIT_FACTORY);
+
+    // Get the Hive table path for the InputSplit created. It is just to satisfy hive. The InputFormat never uses it.
```
Comment (jwang47) [inline cdap-explore/src/main/java/co/cask/cdap/hive/stream/HiveStreamInputFormat.java]: Why the change? Don't fully understand the comment.

Comment (jwang47) [inline cdap-explore/src/main/java/co/cask/cdap/hive/stream/HiveStreamInputFormat.java]: Oh, nevermind saw the StreamInputSplit constructor..

File: cdap-explore/src/main/java/co/cask/cdap/hive/stream/StreamInputSplit.java [hunk 3]
```diff
@@ -45,7 +46,8 @@ public StreamInputSplit() {
   /**
    * Constructs a split.
    *
-   * @param path Path for the stream event file.
+   * @param path the hive file path. Never used by {@link HiveStreamInputFormat}; it is just to satisfy hive.
+   * @param eventPath Path for the stream event file.
```
Comment (jwang47) [inline cdap-explore/src/main/java/co/cask/cdap/hive/stream/StreamInputSplit.java]: Why do we want eventPath and not pass that to super()?

File: cdap-explore/src/main/java/co/cask/cdap/explore/service/ExploreTableManager.java (hunk 1 above)
Comment (chtyim) [inline cdap-explore/src/main/java/co/cask/cdap/explore/service/ExploreTableManager.java]: It was checking for equals, which is the same as "==" (in fact equals check with have a warning).

File: cdap-explore/src/main/java/co/cask/cdap/hive/stream/HiveStreamInputFormat.java (hunk 2 above)
Comment (chtyim) [inline cdap-explore/src/main/java/co/cask/cdap/hive/stream/HiveStreamInputFormat.java]: Updated the javadoc for StreamInputSplit to include the detail of how *funny* Hive behaves.

File: cdap-explore/src/main/java/co/cask/cdap/explore/service/ExploreTableManager.java (hunk 1 above)
Comment (chtyim) [inline cdap-explore/src/main/java/co/cask/cdap/explore/service/ExploreTableManager.java]: Not intended to change the logic, but just to eliminate the warning. I am not sure if the current logic works for sub-class of StructuredRecord.

File: cdap-explore/src/main/java/co/cask/cdap/hive/stream/StreamInputSplit.java (hunk 3 above)
Comment (chtyim) [inline cdap-explore/src/main/java/co/cask/cdap/hive/stream/StreamInputSplit.java]: See Javadoc of this class (may need to reload to see it).

**Comment by jwang47** [comment]
LGTM

**Comment by chtyim** [comment]
Build passed: http://builds.cask.co/browse/CDAP-DUT2677-1
